### PR TITLE
Split tests into different groups

### PR DESCRIPTION
**Pull request by im-denisenko**
Was discussed in https://github.com/ruflin/Elastica/issues/832#issuecomment-101819800.

This PR adds one of *unit*, *functional* or *shutdown* group to every existing test.
It has lot of changes, because almost every setUp method was rewrited to avoid hit ES in tests from *unit* group.
Also, it has some random cleanups like "replace multiple calls of addDocument to one addDocuments".

Now travis will run *functional* tests first, then *shutdown*, and then *unit*. 
It means that unit test can't make any request, because ES cluster is down at the time they runned.

Need to say, I'm quite disappointed about speed increase. I was hoping it'll be more than one minute.

@ruflin review please?

### CODE REVIEW

**Comment by im-denisenko** [comment]
@coveralls you're kidding :)

File: .travis.yml [hunk 1]
```diff
@@ -24,16 +24,17 @@ install:
   - /bin/bash ansible/provision.sh
 
 before_script:
-  - mkdir -p build/logs
+  - mkdir -p build/logs build/coverage
```
Comment (ruflin) [inline .travis.yml]: Based on the new Makefile I suggest we "simplify" the travis.yml file by directly using the make commands (https://github.com/ruflin/Elastica/pull/852). Here for example make prepare

File: .travis.yml [hunk 2]
```diff
@@ -24,16 +24,17 @@ install:
...
   - sh -c 'if [ "$TRAVIS_PHP_VERSION" != "hhvm" ] && [ "$TRAVIS_PHP_VERSION" != "7" ]; then echo "extension=memcache.so" >> ~/.phpenv/versions/$(phpenv version-name)/etc/php.ini; fi;'
   - if [ $ES_COMPOSER_NODEV == "no" ] && ! $(phpenv version-name | grep -q '5.3'); then sudo composer require "guzzlehttp/guzzle" "4.2.*" --dev --no-ansi --no-progress --no-interaction; fi
 
 script:
-  - phpunit -c test/phpunit-travis.xml
+  - phpunit -c test/ --coverage-clover build/coverage/clover-functional.xml --group functional
```
Comment (ruflin) [inline .travis.yml]: make phpunit

File: .travis.yml [hunk 3]
```diff
@@ -24,16 +24,17 @@ install:
...
+  - phpunit -c test/ --coverage-clover build/coverage/clover-shutdown.xml --group shutdown
+  - phpunit -c test/ --coverage-clover build/coverage/clover-unit.xml --group unit
 
 after_script:
   - cat /var/log/elasticsearch/*.log
   - cat /var/log/nginx/*.log
-  - cat build/logs/phpunit-tap.log
-  - sudo composer require satooshi/php-coveralls --no-ansi --no-progress --no-interaction
-  - php vendor/bin/coveralls -v
+  - sudo rm composer.lock && sudo composer require satooshi/php-coveralls dev-master --no-ansi --no-progress --no-interaction
```
Comment (ruflin) [inline .travis.yml]: * is it necessary to remove the .lock file?
* Does the phpunit output still exist?

File: test/lib/Elastica/Test/Base.php [hunk 4]
```diff
@@ -89,10 +89,42 @@ protected function _waitForAllocation(Index $index)
...
+    {
+        parent::setUp();
+
+        $hasGroup = $this->_isUnitGroup() || $this->_isFunctionalGroup() || $this->_isShutdownGroup();
+        $this->assertTrue($hasGroup, 'Every test must have one of "unit", "functional" or "shutdown" group');
+    }
+
     protected function tearDown()
     {
+        if ($this->_isFunctionalGroup()) {
```
Comment (ruflin) [inline test/lib/Elastica/Test/Base.php]: That is cool :+1:

**Comment by ruflin** [comment]
Looks good to me. I see you also did some cleaning up. Thx.

See my comments above.

**Comment by ruflin** [comment]
@im-denisenko I suggest to merge this and make improvements in a second step. About the speed up: The nice part is now that the tests can be split up and only parts can be run. For example for each commit unit tests could be run, but only pull requests all tests are run ... Unfortunately I don't know if such configuration is possible with travis.

File: .travis.yml (hunk 3 above)
Comment (im-denisenko) [inline .travis.yml]: * composer.lock causes composer error when ES_COMPOSER_NODEV is "yes". [Like this](https://travis-ci.org/im-denisenko/Elastica/jobs/62861662#L2177).
* Yep, I remove it. Now phpunit has verbose enough configuration, not like it was in https://github.com/ruflin/Elastica/commit/784049df4dde462ccea45c81f97010ed36191c4a

**Comment by im-denisenko** [comment]
> Unfortunately I don't know if such configuration is possible with travis.

Travis has environment variable TRAVIS_PULL_REQUEST which holds PR id for PR build or false otherwise.

**Comment by ruflin** [comment]
Merged :-)

**Comment by ruflin** [comment]
Interesting. We should then run the unit tests for every commit but functional and shutdown only for pull requests.

**Pull request merged**